utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+where
+    W: Writeable,
+{
+    type W<'a> = &'a W where W: 'a;
+    fn value_for(&self, _key: SinglePlaceholderKey) -> Self::W<'_> {
+        &self.0
+    }
+}
+
+impl<W> PlaceholderValueProvider<SinglePlaceholderKey> for [W; 1]
+where
+    W: Writeable,
+{
+    type W<'a> = &'a W where W: 'a;
+    fn value_for(&self, _key: SinglePlaceholderKey) -> Self::W<'_> {
+        let [value] = self;
+        value
+    }
+}
+
+/// Backend for patterns containing zero or one placeholder.
+///
+/// # Placeholder Keys
+///
+/// The placeholder is always [`SinglePlaceholderKey::Singleton`].
+///
+/// In [`Pattern::interpolate()`], pass a single-element array or tuple.
+///
+/// # Encoding Details
+///
+/// The first code point of the string is 1 plus the byte offset of the placeholder counting from
+/// after that initial code point. If zero, there is no placeholder.
+///
+/// # Examples
+///
+/// Parsing a pattern into the encoding:
+///
+/// ```
+/// use icu_pattern::Pattern;
+/// use icu_pattern::SinglePlaceholder;
+///
+/// // Parse the string syntax and check the resulting data store:
+/// let store = Pattern::<SinglePlaceholder, _>::try_from_str("Hello, {0}!")
+///     .unwrap()
+///     .take_store();
+///
+/// assert_eq!("\u{8}Hello, !", store);
+/// ```
+///
+/// Example patterns supported by this backend:
+///
+/// ```
+/// use icu_pattern::Pattern;
+/// use icu_pattern::SinglePlaceholder;
+///
+/// // Single numeric placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("{0} days ago")
+///         .unwrap()
+///         .interpolate_to_string([5]),
+///     "5 days ago",
+/// );
+///
+/// // Single named placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("{name}")
+///         .unwrap()
+///         .interpolate_to_string(["Alice"]),
+///     "Alice",
+/// );
+///
+/// // No placeholder (note, the placeholder value is never accessed):
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("yesterday")
+///         .unwrap()
+///         .interpolate_to_string(["hi"]),
+///     "yesterday",
+/// );
+///
+/// // Escaped placeholder and a real placeholder:
+/// assert_eq!(
+///     Pattern::<SinglePlaceholder, _>::try_from_str("'{0}' {1}")
+///         .unwrap()
+///         .interpolate_to_string(("hi",)),
+///     "{0} hi",
+/// );
+/// ```
+///
+/// [`Pattern::interpolate()`]: crate::Pattern::interpolate
+#[derive(Debug)]
+pub struct SinglePlaceholder {
+    _not_constructible: core::convert::Infallible,
+}
+
+impl crate::private::Sealed for SinglePlaceholder {}
+
+impl PatternBackend for SinglePlaceholder {
+    type PlaceholderKey = SinglePlaceholderKey;
+    type Store = str;
+    type Iter<'a> = SinglePlaceholderPatternIterator<'a>;
+
+    fn validate_store(store: &Self::Store) -> Result<(), Error> {
+        let placeholder_offset_char = store.chars().next().ok_or(Error::InvalidPattern)?;
+        let initial_offset = char::len_utf8(placeholder_offset_char);
```

Comment:
nit: why fully qualified syntax?
```suggestion
        let initial_offset = placeholder_offset_char.len_utf8();
```

Reply:
There are two other instances of this.